### PR TITLE
[uber-PR] Recent helm-chart improvements plus CI and chart repo publishing

### PR DESCRIPTION
**Pull request by funkypenguin**
Hey guys!

This PR is a combination of my recent PRs, plus some GitHub Actions workflow to publish the helm chart as a repo, using GitHub pages, so that the chart can be installed without cloning the repo. This is particularly useful for those of us who deploy helm charts in CI/CD situations, where cloning a repo and deploying a chart directly from the filesystem is not an option!

This PR incorporates the following discrete PRs:

* #425
* #426 
* #428
* #429
* #430

(I created each PR independently incase you objected to the size / complexity of this one!)

Furthermore, the PR moves the `helm` directory to `charts`, and adds some GitHub workflow to publish a helm repo at https://foundationdb.github.io/fdb-kubernetes-operator/ (_see the update to `README.md`_). The workflow performs a series of linting / testing actions against the helm chart, before publishing it. It's very fussy about the chart version, and requires a semver bump on each change. 

This is an improvement to the workflow I contributed to the k-rail project (https://github.com/cruise-automation/k-rail/tree/k-rail-v2.4.0/.github/workflows).

All that's required to make the helm-repo-publishing magic work, is to add a branch called `gh-pages`, to activate GitHub pages integration.

Cheers!
D

### CODE REVIEW

File: .github/workflows/on-push-lint-charts.yml
Comment (johscheuer) [inline .github/workflows/on-push-lint-charts.yml:23]: What is the reason to install python?

File: .github/workflows/on-push-lint-charts.yml
Comment (johscheuer) [inline .github/workflows/on-push-lint-charts.yml:57]: Missing new line

File: .github/workflows/on-push-lint-charts.yml
Comment (johscheuer) [inline .github/workflows/on-push-lint-charts.yml:29]: I would prefer to move the version into an env variable so we can change it in one place.

File: .github/workflows/on-push-master-publish-chart.yml
Comment (johscheuer) [inline .github/workflows/on-push-master-publish-chart.yml:86]: Missing new line

File: .github/workflows/on-push-master-publish-chart.yml
Comment (johscheuer) [inline .github/workflows/on-push-master-publish-chart.yml:26]: Same like above.

File: .github/workflows/on-push-master-publish-chart.yml
Comment (johscheuer) [inline .github/workflows/on-push-master-publish-chart.yml:3]: Any reason to have this duplicated?

File: Dockerfile
Comment (johscheuer) [inline Dockerfile:50]: Why did you change the UID and GID?

File: README.md
Comment (johscheuer) [inline README.md:16]: I don't see why we should call the helm chart `experimental`?

File: charts/.ci/lint-config.yaml
Comment (johscheuer) [inline charts/.ci/lint-config.yaml:6]: missing new line

File: charts/.ci/scripts/local-ct-lint.sh
Comment (johscheuer) [inline charts/.ci/scripts/local-ct-lint.sh:3]: missing new line

Comment (johscheuer) [reply charts/.ci/scripts/local-ct-lint.sh:3]: Please use `$(pwd)` instead of the deprecated back ticks.

File: charts/.ci/scripts/local-kube-score.sh
Comment (johscheuer) [inline charts/.ci/scripts/local-kube-score.sh:6]: missing new line.

File: charts/.ci/values-kube-score.yaml
Comment (johscheuer) [inline charts/.ci/values-kube-score.yaml:12]: missing new line

File: charts/fdb-operator/Chart.yaml
Comment (johscheuer) [inline charts/fdb-operator/Chart.yaml:17]: This is the chart version and not the `appVersion` so it doesn't have to match the operator version.

File: charts/fdb-operator/templates/manager/deployment.yaml
Comment (johscheuer) [inline charts/fdb-operator/templates/manager/deployment.yaml:25]: Any reason to change the default UID?

File: charts/fdb-operator/templates/rbac/rbac_role_binding.yaml
Comment (johscheuer) [inline charts/fdb-operator/templates/rbac/rbac_role_binding.yaml:24]: missing new line

**Review by johscheuer — CHANGES_REQUESTED** [review]

File: charts/fdb-operator/templates/manager/deployment.yaml
Comment (funkypenguin) [reply charts/fdb-operator/templates/manager/deployment.yaml:25]: Taking the UID above 10000 means that kube-score can validate the **other** components of the securityContext, without complaining about the UID. Apparently UIDs < 10000 are less secure - probably because they could accidentally belong to a system user ;)

File: charts/fdb-operator/Chart.yaml
Comment (funkypenguin) [reply charts/fdb-operator/Chart.yaml:17]: Agreed, and it'll never match, since inevitably changes will need to be made to the chart without changing the operator version. I felt that continuity would be a good place to start though. Happy to change this as required!

File: .github/workflows/on-push-lint-charts.yml
Comment (funkypenguin) [reply .github/workflows/on-push-lint-charts.yml:23]: I don't know :) The chart-testing action seems to require it, it's part of the default example at https://github.com/helm/chart-testing-action

File: README.md
Comment (funkypenguin) [reply README.md:16]: The whole operator is labeled as experimental, but I see your point. No need to re-iterate if we've already stated it.

File: .github/workflows/on-push-master-publish-chart.yml
Comment (funkypenguin) [reply .github/workflows/on-push-master-publish-chart.yml:3]: The idea is that we never publish a chart release if linting has failed. It's technically possible to commit a broken chart, ignore the linting failure, and merge to main. The duplication prevents accidentally publishing broken charts.

File: Dockerfile
Comment (funkypenguin) [reply Dockerfile:50]: As below, to satisfy kube-score, which considers UIDs < 10000 to be a securityContext warning

File: charts/.ci/lint-config.yaml
Comment (funkypenguin) [reply charts/.ci/lint-config.yaml:6]: Ironic that a lint config fails a human lint ;)

File: .github/workflows/on-push-lint-charts.yml
Comment (funkypenguin) [reply .github/workflows/on-push-lint-charts.yml]: Done :)

File: charts/fdb-operator/templates/manager/deployment.yaml
Comment (johscheuer) [reply charts/fdb-operator/templates/manager/deployment.yaml:25]: I would rather vote to leave the old UID I don't see any benefit in changing it and I don't see how UIDs < 10000 should be "less secure". We should rather configure Kube-score to ignore this check: https://github.com/zegl/kube-score/blob/master/README_SECURITYCONTEXT.md#removal-timeline-of-container-security-context

File: .github/workflows/on-push-master-publish-chart.yml
Comment (johscheuer) [reply .github/workflows/on-push-master-publish-chart.yml:3]: I would like to prevent this duplication otherwise we need to maintain multiple files with the same content.

File: .github/workflows/on-push-lint-charts.yml
Comment (johscheuer) [reply .github/workflows/on-push-lint-charts.yml:25]: Could you add a comment that this is a requirement for the chart testing action. They use it for `yamllint` (and some more): Python

**Comment by johscheuer** [comment]
What is your general idea about the PRs and your "uber-PR"? If you don't want to resolve the conflicts on the other PRs and we should only revisit changes in this PR please close them.

**Comment by brownleej** [comment]
I would personally rather have a bunch of smaller PRs :)

**Comment by funkypenguin** [comment]
> What is your general idea about the PRs and your "uber-PR"? If you don't want to resolve the conflicts on the other PRs and we should only revisit changes in this PR please close them.

I'll fix their conflicts - if you merge them first, the resulting diffs from this PR should end up just being the directory rename and the additional CI changes

File: charts/fdb-operator/templates/manager/deployment.yaml
Comment (funkypenguin) [reply charts/fdb-operator/templates/manager/deployment.yaml]: Nice, I didn't know about the improvements to kube-score you linked above. I've updated accordingly, and reset the UID back to default.

File: .github/workflows/on-push-lint-charts.yml
Comment (funkypenguin) [reply .github/workflows/on-push-lint-charts.yml]: Added, and shuffled the order of the steps to make this clearer

**Comment by funkypenguin** [comment]
FYI, I can't fix _these_ conflicts, because I don't have write access to the repo

**Comment by johscheuer** [comment]
You should be able to rebase your branch. I don't see why you need (write) access to this repo to fix the conflicts?

**Comment by funkypenguin** [comment]
Hmm.. I'll give it a try this morning!

**Comment by funkypenguin** [comment]
Rebased and good-to-go! :)

**Comment by funkypenguin** [comment]
This uber-PR has become uber-complicated :( I'm going to close it, wait for my related PRs to merge, and then PR more surgical changes instead.

**Pull request closed without merging**
